File: TP 24 Conteneurisation Spring Angular/Smart_Home_back/src/main/java/com/example/appareil/service/DeviceService.java
package com.example.appareil.service;

import com.example.appareil.entity.Device;
import com.example.appareil.entity.Category;
import com.example.appareil.repository.DeviceRepository;
import com.example.appareil.repository.CategoryRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.rest.webmvc.ResourceNotFoundException;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * Service layer for managing Device operations.
 *
 * <p>
 * This class provides business logic for device management, including
 * CRUD operations and bulk state updates. It ensures devices are
 * associated with valid categories.
 * </p>
 *
 * @author Kelladi fatimaezzahra
 * @version 1.0.0
 * @since 2025-12-29
 */
@Service
public class DeviceService {

    private static final Logger logger = LoggerFactory.getLogger(DeviceService.class);

    @Autowired
    private DeviceRepository deviceRepository;

    @Autowired
    private CategoryRepository categoryRepository;

    /**
     * Saves a new device or updates an existing one.
     *
     * @param device The device to save
     * @return The saved device
     * @throws RuntimeException If the associated category is not found
     */
    public Device save(Device device) {
        logger.info("Saving device: {}", device.getLabel());
        Optional<Category> category = categoryRepository.findById(device.getCategory().getId());
        if (category.isEmpty()) {
            logger.error("Category not found for device: {}", device.getLabel());
            throw new RuntimeException("Category not found.");
        }
        device.setCategory(category.get());
        return deviceRepository.save(device);
    }

    /**
     * Retrieves a device by its identifier.
     *
     * @param id Identifier of the device
     * @return An Optional containing the device if found
     */
    public Optional<Device> findById(Long id) {
        logger.info("Fetching device with ID: {}", id);
        return deviceRepository.findById(id);
    }

    /**
     * Deletes a device by its identifier.
     *
     * @param id Identifier of the device to delete
     */
    public void deleteById(Long id) {
        logger.info("Deleting device with ID: {}", id);
        deviceRepository.deleteById(id);
    }

    /**
     * Retrieves all devices from the database.
     *
     * @return List of all devices
     */
    public List<Device> findAll() {
        logger.info("Fetching all devices from the database.");
        return deviceRepository.findAll();
    }

    /**
     * Updates the state (ON/OFF) of a specific device.
     *
     * @param id      Identifier of the device to update
     * @param devInfo Device information containing the new state
     * @throws ResourceNotFoundException If the device is not found
     */
    public void update(Long id, Device devInfo) {
        logger.info("Updating state for device ID: {}", id);
        Device device = deviceRepository.findById(id)
                .orElseThrow(() -> {
                    logger.error("Device not found with ID: {}", id);
                    return new ResourceNotFoundException("Device not found with id " + id);
                });
        device.setState(devInfo.isState());
        deviceRepository.save(device);
    }

    /**
     * Updates the state of all devices in the system.
     *
     * @param state The new state to apply to all devices
     */
    public void updateAll(boolean state) {
        logger.info("Updating state of all devices to: {}", state ? "ON" : "OFF");
        List<Device> deviceList = deviceRepository.findAll();
        deviceList.forEach(d -> d.setState(state));
        deviceRepository.saveAll(deviceList);
    }
}

File: TP 24 Conteneurisation Spring Angular/smartHome-front/src/app/controller/service/device.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from "@angular/common/http";
import { Observable } from "rxjs";
import { Device } from "../model/Device";
import { Category } from "../model/Category";
import { environment } from "../../../environments/environment";

/**
 * Service for managing device-related operations.
 *
 * <p>This service provides methods to interact with the device and category
 * endpoints of the backend API.</p>
 */
@Injectable({
    providedIn: 'root'
})
export class DeviceService {
    private categoryUrl = environment.URL + "category";
    private deviceUrl = environment.URL + 'device';

    constructor(private http: HttpClient) { }

    /**
     * Retrieves all devices from the backend.
     *
     * @returns Observable of device array
     */
    findAll(): Observable<Device[]> {
        return this.http.get<Device[]>(`${this.deviceUrl}/`);
    }

    /**
     * Retrieves all categories from the backend.
     *
     * @returns Observable of category array
     */
    getCategories(): Observable<Category[]> {
        return this.http.get<Category[]>(`${this.categoryUrl}/`);
    }

    /**
     * Saves a new device to the backend.
     *
     * @param device The device to save
     * @returns Observable of the saved device
     */
    saveDevice(device: Device): Observable<Device> {
        return this.http.post<Device>(`${this.deviceUrl}/`, device);
    }

    /**
     * Deletes a device by its identifier.
     *
     * @param id Identifier of the device to delete
     * @returns Observable of void
     */
    deleteDevice(id: number): Observable<void> {
        return this.http.delete<void>(`${this.deviceUrl}/id/${id}`);
    }

    /**
     * Updates an existing device's information.
     *
     * @param id Identifier of the device to update
     * @param device New device information
     * @returns Observable of the updated device
     */
    updateDevice(id: any, device: any): Observable<Device> {
        return this.http.put<Device>(`${this.deviceUrl}/id/${id}`, device);
    }

    /**
     * Switches the state of all devices in the system.
     *
     * @param state The new state to apply to all devices
     * @returns Observable of void
     */
    switchState(state: boolean): Observable<void> {
        return this.http.get<void>(`${this.deviceUrl}/update/state/${state}`);
    }
}
